compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+from collections.abc import Sequence
+
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.dialects.builtin import IndexType
+from xdsl.ir import Block, Operation, OpResult, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+
+class MoveMemrefAllocations(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving memref allocations outside
+    a double for-loop. This is possible when each loop a new memref-space is allocated with
+    identical sizes and the space is not used outside the loop.
+    """
+
```

Comment:
This is very nice, but this rewritepattern is too complex. I think it makes a lot of sense to split this up in multiple patterns. These can then be combined in one large pass using for example like this:
```python
def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
        PatternRewriteWalker(
            GreedyRewritePatternApplier(
                [
                    Pattern1(),
                    Pattern2(),
                    Pattern3(),
                    ...
                ]
            ),
            apply_recursively=False,
        ).rewrite_module(op)
```
This will then apply all the patterns recursively until nothing changes anymore.
If I'm correct, this pass now applies 3 different patterns:

- If an operation is inside a for loop, and has no side effects (hase the `Pure` trait), and does not depend on anything inside the for loop, it can be hoisted out of the for loop.
- Special case for the dim op -> if it depends on subviews / min, and is only used by allocs, make the approximation, and hoist it out of the for loop.
- Special case for the alloc op -> even though this operation has side effects, we still want to hoist it out of the for loop if it does not depend on anything in the for loop, such that the resulting buffer is reused.